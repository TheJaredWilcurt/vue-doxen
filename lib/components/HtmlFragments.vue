<template>
  <!-- eslint-disable vue/no-v-text-v-html-on-component -->
  <component
    v-for="(node, nodeIndex) in astTree"
    v-bind="node.attributes"
    v-html="node.childrenHtml"
    :is="node.tag"
    :key="'slot-item' + nodeIndex"
  />
</template>

<script>
/**
 * @file This renders HTML markup without a wrapper element (required by v-html).
 */

/* eslint-disable-next-line import/no-extraneous-dependencies */
import { parseDocument } from 'htmlparser2';

export default {
  name: 'HtmlFragments',
  props: {
    html: {
      type: String,
      default: ''
    }
  },
  computed: {
    astTree: function () {
      const xmlOptions = {
        withStartIndices: true,
        withEndIndices: true,
        decodeEntities: false,
        lowerCaseAttributeNames: false,
        normalizeWhitespace: false,
        recognizeSelfClosing: false,
        xmlMode: false
      };
      return parseDocument(this.html, xmlOptions)
        .children
        .filter((node) => {
          if (!node) {
            return false;
          }
          if (node.type === 'tag') {
            return true;
          }
          if (node.type === 'text') {
            const text = this.html.substring(node.startIndex, node.endIndex + 1);
            if (text.trim()) {
              return true
            }
          }
          return false;
        })
        .map((node) => {
          if (node.type === 'text') {
            return {
              attributes: {},
              childrenHtml: this.html.substring(node.startIndex, node.endIndex + 1),
              tag: 'span'
            };
          }
          if (!node.children.length) {
            return {
              attributes: node.attribs,
              childrenHtml: '',
              tag: node.name
            };
          }
          let start = node.endIndex;
          let end = node.startIndex;
          node.children.forEach((child) => {
            if (child.startIndex < start) {
              start = child.startIndex;
            }
            if (child.endIndex > end) {
              end = child.endIndex;
            }
          });
          return {
            attributes: node.attribs,
            childrenHtml: this.html.substring(start, end + 1),
            tag: node.name
          };
        });
    }
  }
};
</script>
