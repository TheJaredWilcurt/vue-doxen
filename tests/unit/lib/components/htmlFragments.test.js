import HtmlFragments from '@/components/HtmlFragments.vue';

import testHelpers from '@@/unit/testHelpers.js';

describe('HtmlFragments.vue', () => {
  beforeEach(() => {
    globalThis.vueSnapshots.removeComments = true;
  });

  const setupWrapper = async (html) => {
    const options = {
      props: {
        html: html || ''
      }
    };
    const wrapper = await testHelpers.mount(HtmlFragments, options);
    return wrapper;
  };

  test('Renders nothing', async () => {
    const wrapper = await setupWrapper();

    expect(wrapper)
      .toMatchSnapshot();
  });

  test('Renders one element', async () => {
    const wrapper = await setupWrapper('<div>Text</div>');

    expect(wrapper)
      .toMatchSnapshot();
  });

  test('Renders one element with children', async () => {
    const wrapper = await setupWrapper([
      '<div class="box">',
      '<span>Text1</span>',
      '<span>Text2</span>',
      '</div>'
    ].join(''));

    expect(wrapper)
      .toMatchSnapshot();
  });


  test('Renders two root elements', async () => {
    const wrapper = await setupWrapper([
      '<div>Text</div>',
      '<p>Paragraph</p>'
    ].join(''));

    expect(wrapper)
      .toMatchSnapshot();
  });

  test('Renders one text node', async () => {
    const wrapper = await setupWrapper('Some Text');

    expect(wrapper)
      .toMatchSnapshot();
  });
});
