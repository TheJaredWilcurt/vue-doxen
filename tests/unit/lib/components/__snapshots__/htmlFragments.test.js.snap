// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HtmlFragments.vue > Renders correctly > Mix of everything 1`] = `
<div></div>
<div
  class="
    box
    cow
  "
  style="background: #F00;"
>
  Content
</div>
<span>
  Text node
</span>
<p class="a">
  <strong class="
    a
    b
    c
  ">
    Text
  </strong>
</p>
`;

exports[`HtmlFragments.vue > Renders correctly > One element with children 1`] = `
<div class="box">
  <span>
    Text1
  </span>
  <span>
    Text2
  </span>
</div>
`;

exports[`HtmlFragments.vue > Renders correctly > One element without children 1`] = `<div></div>`;

exports[`HtmlFragments.vue > Renders correctly > One text node 1`] = `
<span>
  Some Text
</span>
`;

exports[`HtmlFragments.vue > Renders correctly > Partial markup > New wrapper element on previous line 1`] = `
<d div="">
  Text
</d>
`;

exports[`HtmlFragments.vue > Renders correctly > Partial markup > New wrapper element on previous line 2`] = `
<div>
  <div>
    Text
  </div>
</div>
`;

exports[`HtmlFragments.vue > Renders correctly > Partial markup > Type out new wrapper element 1`] = `
<div>
  Text
</div>
`;

exports[`HtmlFragments.vue > Renders correctly > Partial markup > Type out new wrapper element 2`] = `
<div>
  <div>
    Text
  </div>
</div>
`;

exports[`HtmlFragments.vue > Renders correctly > Renders nothing 1`] = ``;

exports[`HtmlFragments.vue > Renders correctly > Two root elements 1`] = `
<div>
  Text
</div>
<p>
  Paragraph
</p>
`;
